cuda/components/warp_blas.cuh

```diff
@@ -354,6 +354,42 @@ __device__ __forceinline__ void multiply_vec(
 }
 
 
+/**
+ * @internal
+ *
+ * This is the same as multiply_vec except this adds the result to res.
+ *
+ * @copydoc multiply_vec
+ */
+template <
+    int max_problem_size, typename Group, typename MatrixValueType,
+    typename VectorValueType,
+    typename = xstd::enable_if_t<group::is_communicator_group<Group>::value>>
+__device__ __forceinline__ void multiply_vec_add(
```

Comment:
If you want to remove the code duplication, you could do something similar to [the reduce function in `ginkgo/cuda/components/reduction.cuh`](https://github.com/ginkgo-project/ginkgo/blob/develop/cuda/components/reduction.cuh#L66), which is using a templated lambda function to define the last step (maybe calling it `Closure closure_op`), and do at the very end:
```
closure_op(res[group.thread_rank() * res_increment] , out);
```

You could make a function that does exactly that, and then call it in both `multiply_vec` and `multiply_vec_add` with a different lambda function. For `multiply_vec_add`, the call would look something like that:
```
multiply_vec_lambda(group, problem_size, vec, mtx_row, mtx_increment, res, res_increment,
                    [](VectorValueType &result, const VectorValueType &out) {result += out;});
```

Reply:
Good idea, I'll do that.

Reply:
Everything should now be fixed, including this. Thanks for the comments.